File: cpp/grid_world.cpp
#include "grid_world.hpp"

GridWorld::GridWorld(const int64_t grid_size)
: grid_size_(grid_size), engine_(std::random_device{}()), dist_(0, grid_size - 1)
{
  self_position_.x = dist_(engine_);
  self_position_.y = dist_(engine_);
  goal_position_.x = dist_(engine_);
  goal_position_.y = dist_(engine_);
}

void GridWorld::print() const
{
  std::cout << "\nstate" << std::endl;
  std::vector<std::vector<int64_t>> v(grid_size_, std::vector<int64_t>(grid_size_, 0));
  v[self_position_.y][self_position_.x] += 1;
  v[goal_position_.y][goal_position_.x] += 2;
  for (int64_t i = 0; i < grid_size_; i++) {
    for (int64_t j = 0; j < grid_size_; j++) {
      std::cout << v[i][j];
    }
    std::cout << std::endl;
  }
}

bool GridWorld::step(const Action action)
{
  if (action == kClick) {
    const bool same = (self_position_ == goal_position_);
    if (same) {
      goal_position_.x = dist_(engine_);
      goal_position_.y = dist_(engine_);
      return true;
    } else {
      return false;
    }
  }

  if (action == kUp) {
    self_position_.y--;
  } else if (action == kRight) {
    self_position_.x++;
  } else if (action == kDown) {
    self_position_.y++;
  } else if (action == kLeft) {
    self_position_.x--;
  }
  self_position_.x = std::clamp(self_position_.x, (int64_t)0, grid_size_ - 1);
  self_position_.y = std::clamp(self_position_.y, (int64_t)0, grid_size_ - 1);
  return false;
}

torch::Tensor GridWorld::state() const
{
  torch::Tensor state = torch::zeros({2, grid_size_, grid_size_});
  state[0][self_position_.y][self_position_.y] = 1;
  state[1][goal_position_.y][goal_position_.y] = 1;
  return state;
}

bool GridWorld::is_ideal_action(const Action action) const
{
  const bool same = (self_position_ == goal_position_);
  if (same) {
    return (action == kClick);
  } else {
    const int64_t dx = goal_position_.x - self_position_.x;
    const int64_t dy = goal_position_.y - self_position_.y;
    if (action == kUp) {
      return (dy < 0);
    } else if (action == kRight) {
      return (dx > 0);
    } else if (action == kDown) {
      return (dy > 0);
    } else if (action == kLeft) {
      return (dx < 0);
    }
  }
}

File: python/grid_world/grid_world.py
import numpy as np
import random
from action import *


class Position:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __eq__(self, other):
        return self.x == other.x and self.y == other.y


class GridWorld:
    def __init__(self, grid_size):
        self.grid_size = grid_size
        self.self_position = Position(random.randint(0, grid_size - 1),
                                      random.randint(0, grid_size - 1))
        self.goal_position = Position(random.randint(0, grid_size - 1),
                                      random.randint(0, grid_size - 1))

    def print(self):
        grid = np.zeros((self.grid_size, self.grid_size), dtype=int)
        grid[self.self_position.y][self.self_position.x] += 1
        grid[self.goal_position.y][self.goal_position.x] += 2
        for row in grid:
            print(''.join(map(str, row)))

    def step(self, action):
        if action == kClick:
            if self.self_position == self.goal_position:
                self.goal_position = Position(random.randint(0, self.grid_size - 1),
                                              random.randint(0, self.grid_size - 1))
                return True
            else:
                return False

        if action == kUp:
            self.self_position.y -= 1
        elif action == kRight:
            self.self_position.x += 1
        elif action == kDown:
            self.self_position.y += 1
        elif action == kLeft:
            self.self_position.x -= 1

        self.self_position.x = max(
            0, min(self.self_position.x, self.grid_size - 1))
        self.self_position.y = max(
            0, min(self.self_position.y, self.grid_size - 1))
        return False

    def state(self):
        state = np.zeros((2, self.grid_size, self.grid_size))
        state[0][self.self_position.y][self.self_position.x] = 1
        state[1][self.goal_position.y][self.goal_position.x] = 1
        return state

    def is_ideal_action(self, action):
        if self.self_position == self.goal_position:
            return action == kClick
        else:
            dx = self.goal_position.x - self.self_position.x
            dy = self.goal_position.y - self.self_position.y
            if action == kUp:
                return dy < 0
            elif action == kRight:
                return dx > 0
            elif action == kDown:
                return dy > 0
            elif action == kLeft:
                return dx < 0
        return False
